File: http-server-netty/src/test/groovy/io/micronaut/docs/http/server/exception/OutOfStockExceptionHandler.java
package io.micronaut.docs.http.server.exception;

import io.micronaut.context.annotation.Requires;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.annotation.Produces;
import io.micronaut.http.server.exceptions.ExceptionHandler;
import javax.inject.Singleton;

@Requires(property = "spec.name", value = "ExceptionHandlerSpec")
//tag::clazz[]
@Produces
@Singleton
@Requires(classes = {OutOfStockException.class, ExceptionHandler.class})
public class OutOfStockExceptionHandler implements ExceptionHandler<OutOfStockException, HttpResponse> {

    @Override
    public HttpResponse handle(HttpRequest request, OutOfStockException exception) {
        return HttpResponse.ok(0);
    }
}
//end::clazz[]

File: test-suite-python/src/test/python/micronaut/docs/server/exception/OutOfStockExceptionHandler.py
import java
from micronaut.context.annotation import Requires
# tag::clazz[]
from jakarta.inject import Singleton
from micronaut.http import HttpRequest, HttpResponse
from micronaut.http.annotation import Produces
from micronaut.http.server.exceptions.response import ErrorContext, ErrorResponseProcessor

from .OutOfStockException import OutOfStockException

ExceptionHandler = java.type("io.micronaut.http.server.exceptions.ExceptionHandler")


# TODO: Re-enable this direct port when Python classes can extend Java Throwable types.
# @Produces
# @Singleton
# @Requires(property="spec.name", value="ExceptionHandlerSpec")
# @Requires(classes=[OutOfStockException, ExceptionHandler])
# class OutOfStockExceptionHandler(ExceptionHandler[OutOfStockException, HttpResponse]):
#
#     def __init__(self, errorResponseProcessor: ErrorResponseProcessor):
#         self.errorResponseProcessor = errorResponseProcessor
#
#     def handle(self, request: HttpRequest, e: OutOfStockException) -> HttpResponse:
#         return self.errorResponseProcessor.processResponse(
#             ErrorContext.builder(request)
#                 .cause(e)
#                 .errorMessage("No stock available")
#                 .build(),
#             HttpResponse.badRequest(),
#         )  # <1>
# end::clazz[]
